rust/src/http2/detect.rs

```diff
@@ -0,0 +1,399 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::http2::{HTTP2FrameTypeData, HTTP2Transaction};
+use super::parser;
+use crate::core::{STREAM_TOCLIENT, STREAM_TOSERVER};
+use std::ffi::CStr;
+use std::mem::transmute;
+use std::str::FromStr;
+
+#[no_mangle]
+pub extern "C" fn rs_http2_tx_get_frametype(
+    tx: *mut std::os::raw::c_void, direction: u8,
+) -> std::os::raw::c_int {
+    let tx = cast_pointer!(tx, HTTP2Transaction);
+    //TODO6tx check use of direction
+    match direction {
+        STREAM_TOSERVER => match &tx.ftype {
+            None => {
+                return -1;
+            }
+            Some(x) => {
+                return *x as i32;
+            }
+        },
+        STREAM_TOCLIENT => match &tx.ftype {
+            None => {
+                return -1;
+            }
+            Some(x) => {
+                return *x as i32;
+            }
+        },
+        _ => {}
+    }
+
+    return -1;
```

Comment:
A the 2 match arms are idential, this could be compressed down to:
```
    match direction {
        STREAM_TOSERVER | STREAM_TOCLIENT => match &tx.ftype {
            None => -1,
            Some(x) => *x as i32
        }
        _ => -1
    }
```

Reply:
Thanks for the Rust tip, I need to get http2 transactions defined right before having a good code here